src/storage/data_table.cpp

```diff
@@ -198,22 +173,25 @@ bool DataTable::IsVisible(const transaction::TransactionContext &txn, const Tupl
 
   // Nullptr in version chain means no version visible to any transaction alive at this point.
   // Alternatively, if the current transaction holds the write lock, it should be able to read its own updates.
-  if (version_ptr == nullptr || version_ptr->Timestamp().load() == txn.TxnId().load()) {
+  if (version_ptr == nullptr || version_ptr->Timestamp().load() == txn->TxnId().load()) {
     return visible;
   }
 
-  // Inspect deltas so we can determine visibility
+  // Apply deltas until we reconstruct a version safe for us to read
+  // If the version chain becomes null, this tuple does not exist for this version, and the last delta
+  // record would be an undo for insert that sets the primary key to null, which is intended behavior.
   while (version_ptr != nullptr &&
-         transaction::TransactionUtil::NewerThan(version_ptr->Timestamp().load(), txn.StartTime())) {
+      transaction::TransactionUtil::NewerThan(version_ptr->Timestamp().load(), txn->StartTime())) {
     // TODO(Matt): It's possible that if we make some guarantees about where in the version chain INSERTs (last position
     // in version chain) and DELETEs (first position in version chain) can appear that we can optimize this check
-    const DeltaRecordType undo_type = StorageUtil::CheckUndoRecordType(*version_ptr);
-    if (undo_type == DeltaRecordType::INSERT) {
-      // Applying the undo of an INSERT makes the tuple invisible to this txn.
-      visible = false;
-    } else if (undo_type == DeltaRecordType::DELETE) {
-      // Applying the undo of a DELETE makes the tuple visible to this txn.
-      visible = true;
+    switch (version_ptr->Type()) {
+      case DeltaRecordType::UPDATE:
+        // Normal delta to be applied. Does not modify the logical delete column.
+        StorageUtil::ApplyDelta(accessor_.GetBlockLayout(), *(version_ptr->Delta()), out_buffer);
+        break;
+      case DeltaRecordType::INSERT:visible = false;
```

Comment:
I think the removed comments might be helpful to a new person looking at the code a couple years from now.